numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -391,8 +391,14 @@ _get_cast_safety_from_castingimpl(PyArrayMethodObject *castingimpl,
   finish:
     Py_DECREF(out_descrs[0]);
     Py_DECREF(out_descrs[1]);
-    /* NPY_NO_CASTING has to be used for (NPY_EQUIV_CASTING|_NPY_CAST_IS_VIEW) */
+    /*
+     * Check for less harmful non-standard returns.  If no cast is necessary,
+     * this must be indicated with `(NPY_NO_CASTING|_NPY_CAST_IS_VIEW)` and
+     * not just `NPY_NO_CASTING`. For casts `NPY_NO_CASTING` is equivalent to:
+     * `(NPY_EQUIV_CASTING|_NPY_CAST_IS_VIEW)` so using that is incorrect.
+     */
     assert(casting != (NPY_EQUIV_CASTING|_NPY_CAST_IS_VIEW));
+    assert(casting != NPY_NO_CASTING);
```

Comment:
I don't understand this comment. It seems you could replace both lines with `(NPY_NO_CASTING |_NPY_CAST_IS_VIEW)`? It is in an `assert` anyway, so is not critical.

Reply:
I added cast-is-view as a flag, because it is explicit and we could also have `NPY_UNSAFE_CASTING | NPY_CAST_IS_VIEW`.  But "no" casting, currently means "equivalent" (at most byte-order should be changing) + "cast-is-view" (since that ensures no byte-order is happening).  So the first checks that "equiv+view" should not be used (it should be "no") and the second checks the "is-view" flag was not forgotten when "no" is returned.

These might not be strict rules in the long term, but I can't think of exceptions right now.

Reply:
I was not checking the logic, rather trying to match the comment to the values. You write: 
> For casts `NPY_NO_CASTING` is equivalent to `(NPY_EQUIV_CASTING|_NPY_CAST_IS_VIEW)` so using that is incorrect

If indeed `NPY_NO_CASTING` = `(NPY_EQUIV_CASTING|_NPY_CAST_IS_VIEW)`, then the two asserts contain exactly the same code. Is there a way to briefly improve the comment to explain why both asserts are needed?

Reply:
OK, I am obviously getting the reasoning not clear.  Changed it to the following, is that clearer?:
```C
    /*
     * Check for less harmful non-standard returns.  The following two returns
     * should never happen. They would be roughly equivalent, but less precise,
     * versions of `(NPY_NO_CASTING|_NPY_CAST_IS_VIEW)`.
     * 1. No-casting must imply cast-is-view.
     * 2. Equivalent-casting + cast-is-view is (currently) the definition
     *    of a "no" cast (there may be reasons to relax this).
     * Note that e.g. `(NPY_UNSAFE_CASTING|_NPY_CAST_IS_VIEW)` is valid.
     */
    assert(casting != NPY_NO_CASTING);
    assert(casting != (NPY_EQUIV_CASTING|_NPY_CAST_IS_VIEW));
```

Reply:
Maybe good to put it in my own words, so that we can check: `NPY_EQUIV_CASTING|_NPY_CAST_IS_VIEW` states that byte-order changes might be needed and that the cast can be done with a view. Since a view implies no byte order changes, this indeed has to be the same as NO_CAST. Should I then interpret your last example similarly: `NPY_UNSAFE_CASTING|_NPY_CAST_IS_VIEW` means unsafe casting may be needed, but it is a view, so it should also be `NPY_NO_CASTING`? Or is this actually meaningful, say, a cast of a float to a structured dtype with just a single float, which can be done as a view?

I guess it has now left me a bit confused about the purpose of the `_NPY_CAST_IS_VIEW` flag - though best may be not to add more here, but instead clarify in `ndarraytypes.h`...

Reply:
> Or is this actually meaningful, say, a cast of a float to a structured dtype with just a single float, which can be done as a view?

Yes.

There are a couple of places where I thought this would be useful. The two main examples are probably:
* A `Float64Unit` dtype is an unsafe cast to `Float64`, but it *is* just a view. (In theory `timedelta64 -> int` falls into this category)
* A structured dtype with different names is just a view, but it should be a "safe" and not an "equiv" cast IMO.

Overall, I feel it is more explicit...  But since I then somewhat enforce these constraints, it might also just be a bit clunky....

In practice, I admit it might be a bit of a half-solution...  There are casts that could be a view+offset and casts that are not "equiv" but can be done as a "view" may just be rare.  And I am not sure that we should try to apply this e.g. to `datetime64_arr.astype(int64, copy=False)` to be a view – although part of that may be that we should check for NaT creation.

Reply:
Its good to discuss this and it is not close to public API yet, so I am happy to reconsider the whole thing :).  For example:

1. We could just drop it again (cast-is-view behaviour is only available for "no" casts).
2. We could make it an output `npy_intp *view_offset` (initialized to INTP_MIN).  That would remove it as a confusing flag on the casting enum and add the flexibility to deal with offsets even.

Reply:
I like having the concept - using views is part of what makes numpy so powerful and I like your extensions. But I think mixing it in with the CAST enum is indeed a bit confusing, and the offset seems a cleaner suggestion (I guess in principle that means one could view an uint64 as an uint32, as long as one did the offset carefully, taking into account bigendian, etc.) 

Perhaps an idea to open a separate issue? Perhaps even a general one with loose ends? For now, it probably makes more sense to proceed with what you have!